[REF] *: Replace underscore functions by native JS (2)

pipu-odoo opened the pull request:
Replace _.isNumber(), _.isArray(), _.filter(), _.reject(), _.unique(), _.indexOf(), _.lastIndexOf(), _.findIndex(), _.range() _.keys(), _.values(), _.str.sprintf() and some _.each()

Description of the refactoring this PR addresses:

Current behavior before PR:
There are underscore.js function enumerated above used in odoo.  

Desired behavior after PR is merged:
These functions has been replaced by native javascript prototypes/methods/functions. 

TaskId : 3246238
---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/118003).

mcm-odoo commented on addons/website/static/src/snippets/s_facebook_page/options.js:
```diff
@@ -19,7 +19,7 @@ options.registry.facebookPage = options.Class.extend({
             hide_cover: true,
         };
-        this.fbData = _.defaults(_.pick(this.$target[0].dataset, _.keys(defaults)), defaults);
+        this.fbData = _.defaults(_.pick(this.$target[0].dataset, Object.keys(defaults)), defaults);
```
import `pick` utils
```suggestion
        this.fbData = Object.assign({}, defaults, pick(this.$target[0].dataset, Object.keys(defaults)), defaults);
```

mcm-odoo commented on addons/web/static/src/legacy/js/widgets/model_field_selector_popover.js:
```diff
@@ -429,7 +430,10 @@ var ModelFieldSelectorPopover = Widget.extend({
 
-        var lines = _.filter(page, this.options.filter);
+
+        //!TODO NEED HELP
+        // this.options.filter is an object like {[arg] : true} 
+        var lines = filter(page, this.options.filter);
```
do not create a new utils for this case as it seems to be the only one.
write a simple for loop to filter the lines.

mcm-odoo commented on addons/web/static/src/legacy/js/widgets/domain_selector.js:
```diff
@@ -868,7 +868,7 @@ var DomainLeaf = DomainNode.extend({
             // Never display "true" or "false" strings from boolean value
             if (_.isBoolean(this.value)) {
```
```suggestion
            if (typeof this.value === "boolean") {
```

pipu-odoo replied:
In the next PR

mcm-odoo commented on addons/web/static/src/legacy/js/widgets/domain_selector.js:
```diff
@@ -868,7 +868,7 @@ var DomainLeaf = DomainNode.extend({
             if (_.isBoolean(this.value)) {
                 this.value = "";
-            } else if (_.isObject(this.value) && !_.isArray(this.value)) { // Can be object if parsed to x2x representation
+            } else if (_.isObject(this.value) && !Array.isArray(this.value)) { // Can be object if parsed to x2x representation
```
```suggestion
            } else if (typeof this.value === "object" && this.value !== null && !Array.isArray(this.value)) { // Can be object if parsed to x2x representation
```

pipu-odoo replied:
In the next PR

pipu-odoo replied:
In the next PR

mcm-odoo commented on addons/web/static/src/legacy/js/views/basic/basic_model.js:
```diff
@@ -2782,12 +2781,14 @@ var BasicModel = AbstractModel.extend({
         }
         return def.then(function (result) {
-            var records = _.uniq(_.flatten(_.values(toFetch)));
+            var records = _.uniq(_.flatten(Object.values(toFetch || {})));
```
looks like `_.flatten` is not necessary
```suggestion
            var records = unique(Object.values(toFetch || {}));
```

pipu-odoo replied:
it seems indeed

mcm-odoo commented on addons/web/static/src/legacy/js/fields/field_utils.js:
```diff
@@ -728,13 +729,13 @@ function parseInteger(value) {
         };
     }
-    if (_.isNumber(value) || _.isString(value)) {
+    if (Number.isFinite(value) || _.isString(value)) {
```
```suggestion
    if (Number.isFinite(value) || typeof value === "string") {
```

pipu-odoo replied:
In the next PR !

mcm-odoo commented on addons/web/static/src/core/utils/objects.js:
```diff
@@ -61,3 +61,20 @@ export function pick(object, ...properties) {
+ * @param {K[]} properties
+ * @returns {T[]} filtered array
+ */
+export function filter(array, properties) {
```
only used once, implement it where it is used

pipu-odoo replied:
now but it could be used elsewhere no ?

mcm-odoo commented on addons/web/static/src/core/utils/numbers.js:
```diff
@@ -32,3 +32,26 @@ export function computeVariation(value, comparisonValue) {
+ */
+export function range(start = 0, stop, step = 1) {
+    const array = [];
+    if (!stop) {
+        stop = start;
+        start = 0;
+    }
```
Better if there is only 1 signature `range(start, stop, step = 1)`

pipu-odoo replied:
ok

mcm-odoo commented on addons/web/static/src/core/utils/numbers.js:
```diff
@@ -32,3 +32,26 @@ export function computeVariation(value, comparisonValue) {
+        stop = start;
+        start = 0;
+    }
+    const nsteps = Math.floor((stop - start) / step);
+    for (let i = 0; i < nsteps; i++) {
+        array.push(start + step * i);
+    }
```
```suggestion
    for (let i = start; i < stop; i += step) {
        array.push(i);
    }
```

pipu-odoo replied:
(>.<)'

pipu-odoo commented on addons/point_of_sale/static/src/js/models.js:
```diff
@@ -1453,7 +1454,7 @@ export class PosGlobalState extends PosModel {
             }
         }
-        return _.uniq(mappedTaxes, (tax) => tax.id);
+        return unique(mappedTaxes, (tax) => tax.id);
```
exact !

pipu-odoo commented on addons/sale/static/src/js/variant_mixin.js:
```diff
@@ -600,7 +601,7 @@ var VariantMixin = {
             precision = parseInt($('.decimal_precision').last().data('precision'));
         }
-        var formatted = _.str.sprintf('%.' + precision + 'f', price).split('.');
+        var formatted = String(Number(price).toFixed(precision)).split(".");
```
ok

pipu-odoo commented on addons/web/static/src/legacy/js/widgets/domain_selector_dialog.js:
argh ... rebase conflict resolution error

pipu-odoo commented on addons/web/static/tests/legacy/helpers/mock_server.js:
```diff
@@ -1870,12 +1866,10 @@ var MockServer = Class.extend({
-        this.data[model].records = _.reject(this.data[model].records, function (record) {
-            return _.contains(ids, record.id);
-        });
+        this.data[model].records = this.data[model].records.filter(record => !_.contains(ids, record.id));
```
Yes i do _.contains in the next PR

pipu-odoo commented on addons/web_editor/static/src/js/common/ace.js:
```diff
@@ -518,9 +519,7 @@ var ViewEditor = Widget.extend({
-            Object.assign(this.views, _.indexBy(_.filter(views, function (view) {
-                return view.active;
-            }), 'id'));
+            Object.assign(this.views, _.indexBy(views.filter(view => view.active), 'id'));
```
_.indexBy() in a next PR

mcm-odoo commented on addons/point_of_sale/static/src/app/navbar/navbar.js:
```diff
@@ -51,10 +51,7 @@ export class Navbar extends Component {
-                            _.str.sprintf(message),
-                            5000
-                        );
+                        this.notification.add(sprintf(message), 5000);
```
this one seems useless, we can give `message` directly without `sprintf`

mcm-odoo commented on addons/survey/static/src/js/survey_session_leaderboard.js:
```diff
@@ -44,7 +45,7 @@ publicWidget.registry.SurveySessionLeaderboard = publicWidget.Widget.extend({
 
         var leaderboardPromise = this._rpc({
-            route: _.str.sprintf('/survey/session/leaderboard/%s', this.surveyAccessToken)
+            route: sprintf('/survey/session/leaderboard/%s', this.surveyAccessToken)
```
```suggestion
            route: `/survey/session/leaderboard/${this.surveyAccessToken}`,
```

mcm-odoo commented on addons/survey/static/src/js/survey_session_leaderboard.js:
```diff
@@ -2,6 +2,7 @@
 
 import publicWidget from "web.public.widget";
 import SESSION_CHART_COLORS from "survey.session_colors";
+import { sprintf } from "@web/core/utils/strings";
```
remove

mcm-odoo commented on addons/survey/static/src/js/survey_form.js:
```diff
@@ -483,7 +484,7 @@ publicWidget.registry.SurveyFormWidget = publicWidget.Widget.extend(SurveyPreloa
 
         var submitPromise = self._rpc({
-            route: _.str.sprintf('%s/%s/%s', route, self.options.surveyToken, self.options.answerToken),
+            route: sprintf('%s/%s/%s', route, self.options.surveyToken, self.options.answerToken),
```
```suggestion
            route: `${route}/${self.options.surveyToken}/${self.options.answerToken}`,
```

pipu-odoo replied:
okok

mcm-odoo commented on addons/survey/static/src/js/survey_form.js:
```diff
@@ -11,6 +11,7 @@ import {getCookie, setCookie, deleteCookie} from "web.utils.cookies";
 
 import SurveyPreloadImageMixin from "survey.preload_image_mixin";
 import { SurveyImageZoomer } from "@survey/js/survey_image_zoomer";
+import { sprintf } from "@web/core/utils/strings";
```
remove

mcm-odoo commented on addons/web/static/src/core/utils/numbers.js:
```diff
@@ -32,3 +32,21 @@ export function computeVariation(value, comparisonValue) {
 }
+
+/**
+ * A function to create flexibly-numbered lists of integers, handy for each and map loops.
+ * start, if omitted, defaults to 0;
+ * step defaults to 1 if start is before stop, otherwise -1.
+ * Returns a list of integers from start (inclusive) to stop (exclusive), incremented (or decremented) by step.
```
comment is outdated

mcm-odoo commented on addons/web/static/src/legacy/js/core/domain.js:
```diff
@@ -299,9 +295,9 @@ var Domain = collections.Tree.extend({
-            throw new Error(_.str.sprintf(
-                "invalid domain %s (missing %d segment(s))",
-                JSON.stringify(domain), expected
+            throw new Error(sprintf(
+                "invalid domain %s (missing %s segment(s))",
+                JSON.stringify(domain), expected.toFixed(0)
             ));
```
```suggestion
            throw new Error(
                `invalid domain ${JSON.stringify(domain)} (missing ${expected.toFixed(0)} segment(s))`);
```

mcm-odoo commented on addons/web/static/src/legacy/js/core/domain.js:
```diff
@@ -1,5 +1,7 @@
 /** @odoo-module alias=web.Domain **/
 
+import { intersection } from "@web/core/utils/arrays";
+import { sprintf } from "@web/core/utils/strings";
```
remove

mcm-odoo commented on addons/website/static/src/js/content/website_root.js:
```diff
@@ -177,7 +178,7 @@ export const WebsiteRoot = publicRootData.PublicRoot.extend(KeyboardNavigationMi
             hash: encodeURIComponent(window.location.hash)
         };
-        window.location.href = _.str.sprintf("/website/lang/%(lang)s?r=%(url)s%(hash)s", redirect);
+        window.location.href = sprintf("/website/lang/%(lang)s?r=%(url)s%(hash)s", redirect);
```
```suggestion
        window.location.href = `/website/lang/${redirect.lang}?r=${redirect.url}${redirect.hash}`;
```

mcm-odoo commented on addons/website/static/src/js/content/website_root.js:
```diff
@@ -7,6 +7,7 @@ import {Markup} from 'web.utils';
 import session from 'web.session';
 import publicRootData from 'web.public.root';
 import "web.zoomodoo";
+import { sprintf } from "@web/core/utils/strings";
```
remove

mcm-odoo commented on addons/website_links/static/src/js/website_links.js:
```diff
@@ -52,7 +53,7 @@ var SelectBox = publicWidget.Widget.extend({
                     return null;
                 }
-                return {id: term, text: _.str.sprintf("Create '%s'", term)};
+                return {id: term, text: sprintf("Create '%s'", term)};
```
```suggestion
                return {id: term, text: `Create '${term}'`};
```

mcm-odoo commented on addons/website_links/static/src/js/website_links.js:
```diff
@@ -1,5 +1,6 @@
 /** @odoo-module alias=website_links.website_links **/
 
+import { sprintf } from "@web/core/utils/strings";
```
remove

mcm-odoo commented on addons/website_slides/static/src/js/slides_course_fullscreen_player.js:
```diff
@@ -644,7 +645,7 @@
                 params.quiz = 1;
             }
-            var fullscreenUrl = _.str.sprintf('%s?%s', url, $.param(params));
+            var fullscreenUrl = sprintf('%s?%s', url, $.param(params));
```
```suggestion
            var fullscreenUrl = `${url}?${$.param(params)}`;
```

mcm-odoo commented on addons/website_slides/static/src/js/slides_course_fullscreen_player.js:
```diff
@@ -12,6 +12,7 @@
     import { SlideCoursePage } from '@website_slides/js/slides_course_page';
     import Dialog from 'web.Dialog';
     import '@website_slides/js/slides_course_join';
+import { sprintf } from '@web/core/utils/strings';
```
remove

mcm-odoo commented on addons/point_of_sale/static/src/js/models.js:
```diff
@@ -1451,7 +1452,7 @@ export class PosGlobalState extends PosModel {
             }
         }
-        return _.uniq(mappedTaxes, (tax) => tax.id);
+        return uniqueBy(mappedTaxes, (tax) => tax.id);
```
Looks like it is used only once. You can move the utils on top of this file after the imports.

mcm-odoo commented on addons/web/static/src/core/utils/arrays.js:
```diff
@@ -203,3 +203,24 @@ export function sections(array) {
+ * @param {F} function
+ * @returns {T[]}
+ */
+export function uniqueBy(array, agg) {
```
to move

pipu-odoo commented:
@odoo/rd-security 

[odoo/addons/mass_mailing/static/src/js/mailing_portal.js:49]
$('#subscription_info').html(sprintf(
=> I changed underscore _.str.sprintf by sprintf utils function (addons/web/static/src/core/utils/strings.js)
=> _.escape() is still used in .html() 

[odoo/addons/survey/static/src/js/survey_session_manage.js:253]
document.location = /survey/results/${self.surveyId};
=> I only changed underscore _.str.sprintf by string

[odoo/addons/website/static/src/js/content/website_root.js:180]
window.location.href = /website/lang/${redirect.lang}?r=${redirect.url}${redirect.hash};
=> I only changed underscore _.str.sprintf by string

[odoo/addons/website/static/src/js/tours/tour_utils.js:108]
content: Markup(sprintf(_t("<b>Click</b> on this option to change the %s of the block."), optionTooltipLabel)),
=> I changed underscore _.str.sprintf by sprintf utils function (addons/web/static/src/core/utils/strings.js)

[odoo/addons/website/static/src/js/tours/tour_utils.js:119]
content: Markup(sprintf(_t("<b>Select</b> a %s."), optionTooltipLabel)),
=> I changed underscore _.str.sprintf by sprintf utils function (addons/web/static/src/core/utils/strings.js)

[odoo/addons/website/static/src/js/tours/tour_utils.js:136]
content: Markup(sprintf(_t("<b>Slide</b> this button to change the %s padding"), direction)),
=> I changed underscore _.str.sprintf by sprintf utils function (addons/web/static/src/core/utils/strings.js)

[odoo/addons/website/static/src/js/tours/tour_utils.js:233]
content: Markup(sprintf(_t("Drag the <b>%s</b> building block and drop it at the bottom of the page."), snippet.name)),
=> I changed underscore _.str.sprintf by sprintf utils function (addons/web/static/src/core/utils/strings.js)

caburj commented on addons/point_of_sale/static/src/js/Screens/ProductScreen/ControlButtons/SaveButton.js:
```diff
@@ -3,44 +3,34 @@
     onClick() {
         const orderline = this.env.pos.get_order().get_selected_orderline();
         if (!orderline) {
-            this.notification.add(
```
you can remove the use of sprintf

caburj commented on addons/point_of_sale/static/src/js/Screens/ProductScreen/ControlButtons/SaveButton.js:
```diff
@@ -3,44 +3,34 @@
             return;
         }
         this._selectEmptyOrder();
-        this.notification.add(
```
you can remove the use of sprintf

caburj commented on addons/point_of_sale/static/src/js/Screens/ProductScreen/ProductScreen.js:
```diff
@@ -241,10 +241,7 @@ export class ProductScreen extends ControlButtonsMixin(Component) {
                     this.pos.showScreen("TicketScreen", {
                         ui: { filter: "SYNCED", searchDetails },
                     });
-                    this.notification.add(
```
you can remove the use of sprintf

pipu-odoo commented on addons/website/static/src/js/content/website_root.js:
```diff
@@ -177,7 +177,7 @@ export const WebsiteRoot = publicRootData.PublicRoot.extend(KeyboardNavigationMi
             hash: encodeURIComponent(window.location.hash)
         };
-        window.location.href = _.str.sprintf("/website/lang/%(lang)s?r=%(url)s%(hash)s", redirect);
+        window.location.href = `/website/lang/${redirect.lang}?r=${redirect.url}${redirect.hash}`;
```
Here it make sens

pipu-odoo commented:
@odoo/rd-security changes has been done regarding your comments.

pipu-odoo commented on addons/website/static/src/js/content/website_root.js:
```diff
@@ -177,7 +178,7 @@ export const WebsiteRoot = publicRootData.PublicRoot.extend(KeyboardNavigationMi
             hash: encodeURIComponent(window.location.hash)
         };
-        window.location.href = _.str.sprintf("/website/lang/%(lang)s?r=%(url)s%(hash)s", redirect);
+        window.location.href = escape(`/website/lang/${redirect.lang}?r=${redirect.url}${redirect.hash}`);
```
Ok

pipu-odoo commented:
@xmo-odoo Ok for you now ?
